widget/tree.go

```diff
@@ -0,0 +1,781 @@
+package widget
+
+import (
+	"image/color"
+	//"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/driver/desktop"
+	"fyne.io/fyne/internal/widget"
+	"fyne.io/fyne/storage"
+	"fyne.io/fyne/theme"
+)
+
+const (
+	indentation = 4 // Multiplier of theme.Padding() to indent children by. // TODO consider moving to theme.Indentation()
+)
+
+var _ fyne.Widget = (*Tree)(nil)
+
+// Tree widget displays hierarchical data.
+// Each node of the tree must be identified by a Unique ID.
+type Tree struct {
+	BaseWidget
+	Root     string
+	Selected string
+	Offset   fyne.Position
+
+	open map[string]bool
+
+	Children       func(uid string) (c []string)                         // Return a sorted slice of Children Unique IDs for the given Node Unique ID
+	IsBranch       func(uid string) (ok bool)                            // Return true if the given Unique ID represents a Branch
+	NewNode        func(branch bool) (o fyne.CanvasObject)               // Return a CanvasObject that can represent a Branch (if branch is true), or a Leaf (if branch is false)
+	UpdateNode     func(uid string, branch bool, node fyne.CanvasObject) // Update the given CanvasObject to represent the data at the given Unique ID
+	OnBranchOpened func(uid string)                                      // Called when a Branch is opened
+	OnBranchClosed func(uid string)                                      // Called when a Branch is closed
+	OnNodeSelected func(uid string, node fyne.CanvasObject)              // Called when the Node with the given Unique ID is selected.
+}
+
+// NewTreeWithStrings creates a new tree with the given string map.
+// Data must contain a mapping for the root, which defaults to empty string ("").
+func NewTreeWithStrings(data map[string][]string) (t *Tree) {
+	t = &Tree{
+		open: make(map[string]bool),
+		Children: func(uid string) (c []string) {
+			c, _ = data[uid]
+			//log.Println("StringTree.Children:", uid, c)
+			return
+		},
+		IsBranch: func(uid string) (b bool) {
+			_, b = data[uid]
+			//log.Println("StringTree.IsBranch:", uid, b)
+			return
+		},
+		NewNode: func(branch bool) (o fyne.CanvasObject) {
+			o = NewLabel("")
+			//log.Println("StringTree.NewNode:", branch, o)
+			return
+		},
+	}
+	t.UpdateNode = func(uid string, branch bool, node fyne.CanvasObject) {
+		//log.Println("StringTree.UpdateNode:", uid, branch, node)
+		l := node.(*Label)
+		l.SetText(uid)
+		//log.Println("StringTree.Label:", l.Text)
+	}
+	t.ExtendBaseWidget(t)
+	return
+}
+
+// NewTreeWithFiles creates a new tree with the given file system URI.
+// TODO contents of a directory are sorted with the given sorter
+func NewTreeWithFiles(root fyne.URI) (t *Tree) {
```

Comment:
It doesn't make much difference implementation-wise as it is immediately converted to a string, which `Children` recreates into a URI, and then Listable URI.

It was originally a string parameter, but it was changed to fyne.URI to indicate it can't be just any string. It could be changed to ListableURI so it is obvious to the dev that it should be listable, but that conveys an incorrect requirement - it doesn't have to be listable, a single leaf tree is still a valid state.